### PR TITLE
Feat/favorite

### PR DESCRIPTION
**Pull request by FirwoodLin**
#### Description :

- Completed the basic functionality of the favorite module
- Awaiting the addition of API by the feed module
- Unit tests are to be added(In Progress)

### CODE REVIEW

**Comment by Dyamidsteve** [comment]
wg.Add(1)
go func() {
	defer wg.Done()
	// connect to redis......
}()
wg.Wait()

In such a code,it cost additional and unneccesary performance loss to start another groutine,because your main goroutine doesn't do anything but wait for the connection with redis.You should use goroutine in a more feasible way.

**Comment by intyouss** [comment]
> wg.Add(1)
> go func() {
> 	defer wg.Done()
> 	// connect to redis......
> }()
> wg.Wait()
> 
> In such a code,it cost additional and unneccesary performance loss to start another groutine,because your main goroutine doesn't do anything but wait for the connection with redis.You should use goroutine in a more feasible way. 

The Clearup method is to close the connection between the redis server and the mysql server, not to let the main thread continue to execute, because this method is the cleanup method called after the main thread has ended，no proplem.

**Comment by intyouss** [comment]
The NewRedisConn method does not necessarily add goroutine, and the coroutine you add blocks it inside the method, which is invalid, equivalent to serial execution, and increases

**Comment by FirwoodLin** [comment]
seems like I have to modify the porto on the feat/favorite and open another PR for the sake of too many conflicts in proto and generated file

**Comment by FirwoodLin** [comment]
> seems like I have to modify the porto on the feat/favorite and open another PR for the sake of too many conflicts in proto and generated file

fixed

File: app/favorite/service/internal/biz/favorite.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		return err
+	}
+	if len(videoInfo) == 0 {
+		return errors.New("video not found")
+	}
+	authorId := videoInfo[0].Author.Id // to update author's favorited
+	switch actionType {
+	case 1:
+		return uc.tx.ExecTx(ctx, func(ctx context.Context) error {
```
Comment (intyouss) [inline app/favorite/service/internal/biz/favorite.go]: The biz tier should be business processing without any functionality related to database interaction, your transaction processing is out of bounds, please put him in the data tier

File: app/favorite/service/internal/data/data.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		go func() {
+			defer wg.Done()
+			sqlDB, err := db.DB()
+			// 如果err不为空，则连接池中没有连接
+			if err != nil {
+				return
+			}
+			if err = sqlDB.Close(); err != nil {
+				logHelper.Errorf("Mysql connection closure failed, err: %w", err)
+				return
```
Comment (intyouss) [inline app/favorite/service/internal/data/data.go]: In the clearup method, the comment service has tested the method to close the gorm connection pool. Shutting down mysql using this method has a bug: it takes two shutdowns to actually close the connection. The explanation given is that Gorm has its own set of methods for automatically closing connections. There is no need for the user to close it manually. So in the clearup method, please discard the manual close method. Clearup bug

File: app/favorite/service/internal/biz/favorite.go (hunk 1 above)
Comment (FirwoodLin) [reply app/favorite/service/internal/biz/favorite.go]: ok, i'll fix it

File: app/favorite/service/internal/data/favorite.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	"time"
+)
+
+// Favorite Database Model
+type Favorite struct {
+	ID        uint32     `gorm:"column:id;primary_key;autoIncrement"`
+	VideoID   uint32     `gorm:"column:video_id"`
+	UserID    uint32     `gorm:"column:user_id"`
+	CreatedAt time.Time  `gorm:"column:created_at"` // new add field; for backend use only
+	DeletedAt *time.Time `gorm:"column:deleted_at;index"`
```
Comment (intyouss) [inline app/favorite/service/internal/data/favorite.go]: Your delete time doesn't use gorm.DeleteAt. can't perform gorm's soft delete

File: app/favorite/service/internal/data/data.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+			if err = cacheClient.favoriteCache.client.Close(); err != nil {
+				logHelper.Errorf("Redis connection closure failed, err: %w", err)
+			}
+			logHelper.Info("Successfully close the Redis connection")
+		}()
+		wg.Wait()
+	}
+
+	data := &Data{
+		db:    db,
```
Comment (intyouss) [inline app/favorite/service/internal/data/data.go]: You should set the table model in advance in your database connection. db.Model(&Favorite{}), which prevents some unnecessary bugs and speeds up the database locking table.

File: app/favorite/service/internal/server/server.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+type PublishConn stdgrpc.ClientConnInterface
+
+// NewUserClient 创建一个User服务客户端，接收User服务数据
+func NewUserClient(c *conf.Client, logger log.Logger) UserConn {
+	conn, err := grpc.DialInsecure(
+		context.Background(),
+		grpc.WithEndpoint(c.User.To),
+		grpc.WithMiddleware(
+			recovery.Recovery(),
+			logging.Server(logger),
```
Comment (intyouss) [inline app/favorite/service/internal/server/server.go]: Fix bug. change logging.Server to logging.Client.

File: app/favorite/service/internal/server/server.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+// NewPublishClient 创建一个Publish服务客户端，接收Publish服务数据
+func NewPublishClient(c *conf.Client, logger log.Logger) PublishConn {
+	conn, err := grpc.DialInsecure(
+		context.Background(),
+		grpc.WithEndpoint(c.Publish.To),
+		grpc.WithMiddleware(
+			recovery.Recovery(),
+			logging.Server(logger),
```
Comment (intyouss) [inline app/favorite/service/internal/server/server.go]: This one's the same.

**Review by intyouss — COMMENTED** [review]
Overall, it's great. Just one thing next time, the modifications will be merged

File: api/favorite/service/v1/favorite.proto [hunk 7]
```diff
@@ -1,53 +1,62 @@
...
-	rpc GetTotalFavorited(TotalFavoritedRequest) returns (TotalFavoritedReply) {}	// get favorited of user(user received)
-	rpc GetVideoFavoriteCount(VideoFavoriteCountRequest) returns (VideoFavoriteCountReply) {}	// get favorited of video
-	rpc GetUserFavoriteCount(UserFavoriteCountRequest) returns (UserFavoriteCountReply) {}	// get favorite count of user(user sent)
+	// 根据userId和videoId判断是否喜爱(publish)
+	rpc IsFavorite(IsFavoriteRequest) returns (IsFavoriteReply) {}
+}
+
+message IsFavoriteRequest {
+	uint32 user_id = 1; // 用户id
+	uint32 video_id = 3; // 视频id
```
Comment (intyouss) [inline api/favorite/service/v1/favorite.proto]: Please modify to be repeated uint32 video_ids

File: api/favorite/service/v1/favorite.proto [hunk 8]
```diff
@@ -1,53 +1,62 @@
...
+}
+
+message IsFavoriteReply {
+	int32 status_code = 1; // 状态码，0-成功，其他值-失败
+	string status_msg = 2; // 返回状态描述
+	bool is_favorite = 3; // true-已点赞，false-未点赞
```
Comment (intyouss) [inline api/favorite/service/v1/favorite.proto]: repeated bool is_favorite

**Review by intyouss — APPROVED** [review]
LGTM,Great attitude, great modification

**Review by mo3et — APPROVED** [review]
LGTM.

**Pull request merged**